File: xsolla-login-sdk/src/main/java/com/xsolla/android/login/api/LoginApi.java
package com.xsolla.android.login.api;

import com.xsolla.android.login.entity.common.UserAttribute;
import com.xsolla.android.login.entity.request.AuthUserBody;
import com.xsolla.android.login.entity.request.AuthUserSocialBody;
import com.xsolla.android.login.entity.request.GetUsersAttributesFromClientRequest;
import com.xsolla.android.login.entity.request.OauthAuthUserBody;
import com.xsolla.android.login.entity.request.OauthGetCodeBySocialTokenBody;
import com.xsolla.android.login.entity.request.OauthRegisterUserBody;
import com.xsolla.android.login.entity.request.RegisterUserBody;
import com.xsolla.android.login.entity.request.ResetPasswordBody;
import com.xsolla.android.login.entity.request.UpdateUserDetailsBody;
import com.xsolla.android.login.entity.request.UpdateUserFriendsRequest;
import com.xsolla.android.login.entity.request.UpdateUserPhoneBody;
import com.xsolla.android.login.entity.request.UpdateUsersAttributesFromClientRequest;
import com.xsolla.android.login.entity.request.UpdateUsersAttributesFromServerRequest;
import com.xsolla.android.login.entity.response.AuthResponse;
import com.xsolla.android.login.entity.response.AuthSocialResponse;
import com.xsolla.android.login.entity.response.LinkForSocialAuthResponse;
import com.xsolla.android.login.entity.response.OauthAuthResponse;
import com.xsolla.android.login.entity.response.OauthGetCodeBySocialTokenResponse;
import com.xsolla.android.login.entity.response.OauthLinkForSocialAuthResponse;
import com.xsolla.android.login.entity.response.SearchUsersByNicknameResponse;
import com.xsolla.android.login.entity.response.SocialFriendsResponse;
import com.xsolla.android.login.entity.response.UserDetailsResponse;
import com.xsolla.android.login.entity.response.UserFriendsResponse;
import com.xsolla.android.login.entity.response.UserPublicInfoResponse;

import java.util.List;

import okhttp3.MultipartBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Multipart;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface LoginApi {

    // JWT

    @POST("/api/user")
    Call<Void> registerUser(
            @Query("projectId") String projectId,
            @Body RegisterUserBody registerUserBody
    );

    @POST("/api/login")
    Call<AuthResponse> login(
            @Query("projectId") String projectId,
            @Query("with_logout") String withLogout,
            @Body AuthUserBody authUserBody
    );

    @POST("/api/social/{providerName}/login_with_token")
    Call<AuthSocialResponse> loginSocial(
            @Path("providerName") String providerName,
            @Query("projectId") String projectId,
            @Query("with_logout") String withLogout,
            @Body AuthUserSocialBody authUserSocialBody
    );


    // Profile and friends

    @GET("/api/social/{providerName}/login_url")
    Call<LinkForSocialAuthResponse> getLinkForSocialAuth(
            @Path("providerName") String providerName,
            @Query("projectId") String projectId,
            @Query("with_logout") String withLogout
    );

    @GET("/api/users/me/social_friends")
    Call<SocialFriendsResponse> getSocialFriends(
            @Header("authorization") String authHeader,
            @Query("platform") String platform,
            @Query("offset") int offset,
            @Query("limit") int limit,
            @Query("with_xl_uid") boolean fromGameOnly
    );

    @GET("api/users/search/by_nickname")
    Call<SearchUsersByNicknameResponse> searchUsersByNickname(
            @Header("authorization") String authHeader,
            @Query("nickname") String nickname,
            @Query("offset") int offset,
            @Query("limit") int limit
    );

    @GET("api/users/{userId}/public")
    Call<UserPublicInfoResponse> getUserPublicInfo(
            @Header("authorization") String authHeader,
            @Path("userId") String userId
    );

    @GET("api/users/me")
    Call<UserDetailsResponse> getCurrentUserDetails(
            @Header("authorization") String authHeader
    );

    @PATCH("api/users/me")
    Call<Void> updateCurrentUserDetails(
            @Header("authorization") String authHeader,
            @Body UpdateUserDetailsBody updateUserDetailsBody
    );

    @DELETE("api/users/me/picture")
    Call<Void> deleteUserPicture(
            @Header("authorization") String authHeader
    );

    @Multipart
    @POST("api/users/me/picture")
    Call<Void> uploadUserPicture(
            @Header("authorization") String authHeader,
            @Part MultipartBody.Part filePart
    );

    @POST("api/users/me/phone")
    Call<Void> updateUserPhone(
            @Header("authorization") String authHeader,
            @Body UpdateUserPhoneBody updateUserPhoneBody
    );

    @DELETE("api/users/me/phone/{phoneNumber}")
    Call<Void> deleteUserPhone(
            @Header("authorization") String authHeader,
            @Path("phoneNumber") String phoneNumber
    );

    @GET("api/users/me/relationships")
    Call<UserFriendsResponse> getUserFriends(
            @Header("authorization") String authHeader,
            @Query("after") String after,
            @Query("limit") int limit,
            @Query("type") String userFriendsRequestType,
            @Query("sort_by") String userFriendsRequestSortBy,
            @Query("sort_order") String userFriendsRequestSortOrder
    );

    @POST("api/users/me/relationships")
    Call<Void> updateFriends(
            @Header("authorization") String authHeader,
            @Body UpdateUserFriendsRequest updateUserFriendsRequest
    );

    @POST("api/attributes/users/me/get")
    Call<List<UserAttribute>> getUsersAttributesFromClient(
            @Header("authorization") String authHeader,
            @Body GetUsersAttributesFromClientRequest getUsersAttributesFromClientRequest
    );

    @POST("api/attributes/users/me/get_read_only")
    Call<List<UserAttribute>> getUsersReadOnlyAttributesFromClient(
            @Header("authorization") String authHeader,
            @Body GetUsersAttributesFromClientRequest getUsersAttributesFromClientRequest
    );

    @POST("api/attributes/users/me/update")
    Call<Void> updateUsersAttributesFromClient(
            @Header("authorization") String authHeader,
            @Body UpdateUsersAttributesFromClientRequest updateUsersAttributesFromClientRequest
    );

    @POST("api/attributes/users/{userId}/update")
    Call<Void> updateUsersAttributesFromServer(
            @Path("userId") String userId,
            @Body UpdateUsersAttributesFromServerRequest updateUsersAttributesFromServerRequest
    );

    @POST("api/attributes/users/{userId}/update_read_only")
    Call<Void> updateUsersReadOnlyAttributesFromServer(
            @Path("userId") String userId,
            @Body UpdateUsersAttributesFromServerRequest updateUsersAttributesFromServerRequest
    );


    // OAuth 2.0

    @POST("/api/oauth2/user")
    Call<Void> oauthRegisterUser(
            @Query("response_type") String responseType,
            @Query("client_id") int clientId,
            @Query("scope") String scope,
            @Query("state") String state,
            @Query("redirect_uri") String redirectUri,
            @Body OauthRegisterUserBody body
    );

    @POST("/api/oauth2/login/token")
    Call<OauthAuthResponse> oauthLogin(
            @Query("client_id") int clientId,
            @Query("scope") String scope,
            @Body OauthAuthUserBody oauthAuthUserBody
    );

    @FormUrlEncoded
    @POST("/api/oauth2/token")
    Call<OauthAuthResponse> oauthRefreshToken(
            @Field("refresh_token") String refreshToken,
            @Field("grant_type") String grantType,
            @Field("client_id") int clientId,
            @Field("redirect_uri") String redirectUri
    );

    @FormUrlEncoded
    @POST("/api/oauth2/token")
    Call<OauthAuthResponse> oauthGetTokenByCode(
            @Field("code") String code,
            @Field("grant_type") String grantType,
            @Field("client_id") int clientId,
            @Field("redirect_uri") String redirectUri
    );

    @GET("/api/oauth2/social/{providerName}/login_url")
    Call<OauthLinkForSocialAuthResponse> oauthGetLinkForSocialAuth(
            @Path("providerName") String providerName,
            @Query("client_id") int clientId,
            @Query("state") String state,
            @Query("redirect_uri") String redirectUri,
            @Query("response_type") String responseType,
            @Query("scope") String scope
    );

    @POST("/api/oauth2/social/{providerName}/login_with_token")
    Call<OauthGetCodeBySocialTokenResponse> oauthGetCodeBySocialToken(
            @Path("providerName") String providerName,
            @Query("client_id") int clientId,
            @Query("state") String state,
            @Query("redirect_uri") String redirectUri,
            @Query("response_type") String responseType,
            @Query("scope") String scope,
            @Body OauthGetCodeBySocialTokenBody oauthGetCodeBySocialTokenBody
    );


    // General

    @POST("/api/password/reset/request")
    Call<Void> resetPassword(
            @Query("projectId") String projectId,
            @Body ResetPasswordBody resetPasswordBody
    );
}

File: xsolla-login-sdk/src/main/java/com/xsolla/android/login/api/LoginApi.kt
package com.xsolla.android.login.api

import com.xsolla.android.login.entity.common.UserAttribute
import com.xsolla.android.login.entity.request.*
import com.xsolla.android.login.entity.response.*
import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*

interface LoginApi {
    //----------    Authentication      ----------

    //
    // Authentication
    //
    // JWT

    @POST("/api/login")
    fun login(
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String?,
        @Query("payload") payload: String?,
        @Query("with_logout") withLogout: String,
        @Body authUserBody: AuthUserBody
    ): Call<AuthResponse>

    @POST("/api/login/device/{device_type}")
    fun authViaDeviceId(
        @Path("device_type") deviceType: String,
        @Query("projectId") projectId: String,
        @Query("payload") payload: String?,
        @Query("with_logout") withLogout: String,
        @Body loginByDeviceIdBody: AuthViaDeviceIdBody
    ) : Call <AuthViaIdResponse>

    @POST("api/login/phone/confirm")
    fun completeAuthByPhone(
        @Query("projectId") projectId: String,
        @Body completeAuthByPhoneBody: CompleteAuthByPhoneBody
    ): Call<AuthResponse>

    @POST("api/login/phone/request")
    fun startAuthByPhone(
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String,
        @Query("payload") payload: String?,
        @Query("with_logout") withLogout: String,
        @Body startAuthByPhoneBody: StartAuthByPhoneBody
    ): Call<StartAuthByMobileResponse>

    @GET("/api/social/{providerName}/login_url")
    fun getLinkForSocialAuth(
        @Path("providerName") providerName: String,
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String?,
        @Query("fields") fields: List<String>?,
        @Query("payload") payload: String?,
        @Query("with_logout") withLogout: String
    ): Call<LinkForSocialAuthResponse>

    @POST("/api/social/{providerName}/login_with_token")
    fun authViaAccessTokenOfSocialNetwork(
        @Path("providerName") providerName: String,
        @Query("projectId") projectId: String,
        @Query("payload") payload: String?,
        @Query("with_logout") withLogout: String,
        @Body authUserSocialBody: AuthUserSocialBody
    ): Call<AuthSocialResponse>

    @POST("/api/user")
    fun registerUser(
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String?,
        @Query("payload") payload: String?,
        @Body registerUserBody: RegisterUserBody
    ): Call<Void>


    //Authentication
    //
    //OAuth 2.0


    @POST("/api/oauth2/login/token")
    fun oauthLogin(
        @Query("client_id") clientId: Int,
        @Query("scope") scope: String,
        @Body oauthAuthUserBody: OauthAuthUserBody
    ): Call<OauthAuthResponse>

    @POST("/api/oauth2/login/device/{device_type}")
    fun oauthAuthViaDeviceId(
        @Path("device_type") deviceType: String,
        @Query("client_id") client:Int,
        @Query("response_type") responseType: String,
        @Query("redirect_uri") redirectUri: String,
        @Query("state") state: String,
        @Query("scope") scope: String,
        @Body body: AuthViaDeviceIdBody
    ) : Call<OauthGetCodeResponse>

    @POST("api/oauth2/login/phone/confirm")
    fun oauthCompleteAuthByPhone(
        @Query("client_id") clientId: Int,
        @Body completeByPhoneBody: CompleteAuthByPhoneBody
    ): Call<OauthAuthResponse>

    @POST("api/oauth2/login/phone/request")
    fun oauthStartAuthByPhone(
        @Query("response_type") responseType: String,
        @Query("client_id") clientId: Int,
        @Query("scope") scope: String,
        @Query("state") state: String,
        @Query("redirect_uri") redirectUri: String,
        @Body body: StartAuthByPhoneBody
    ): Call<StartAuthByMobileResponse>

    @GET("/api/oauth2/social/{providerName}/login_url")
    fun oauthGetLinkForSocialAuth(
        @Path("providerName") providerName: String,
        @Query("client_id") clientId: Int,
        @Query("state") state: String,
        @Query("redirect_uri") redirectUri: String,
        @Query("response_type") responseType: String,
        @Query("scope") scope: String
    ): Call<OauthLinkForSocialAuthResponse>

    @POST("/api/oauth2/social/{providerName}/login_with_token")
    fun oauthGetCodeBySocialToken(
        @Path("providerName") providerName: String,
        @Query("client_id") clientId: Int,
        @Query("state") state: String,
        @Query("redirect_uri") redirectUri: String,
        @Query("response_type") responseType: String,
        @Query("scope") scope: String,
        @Body oauthGetCodeBySocialTokenBody: OauthGetCodeBySocialTokenBody
    ): Call<OauthGetCodeResponse>

    @FormUrlEncoded
    @POST("/api/oauth2/token")
    fun oauthGetTokenByCode(
        @Field("code") code: String,
        @Field("grant_type") grantType: String,
        @Field("client_id") clientId: Int,
        @Field("redirect_uri") redirectUri: String
    ): Call<OauthAuthResponse>

    @POST("/api/oauth2/user")
    fun oauthRegisterUser(
        @Query("response_type") responseType: String,
        @Query("client_id") clientId: Int,
        @Query("scope") scope: String,
        @Query("state") state: String,
        @Query("redirect_uri") redirectUri: String,
        @Body body: OauthRegisterUserBody
    ): Call<Void>

    @POST("/api/social/mobile/{providerName}/login_with_code")
    fun loginSocialWithOauthCode(
        @Path("providerName") providerName: String,
        @Query("projectId") projectId: String,
        @Query("with_logout") withLogout: String,
        @Body authUserSocialWithCodeBody: AuthUserSocialWithCodeBody
    ): Call<AuthSocialResponse>

    //----------     Emails     ----------


    // Emails
    //
    // JWT

    @POST("/user/resend_confirmation_link")
    fun resendAccountConfirmationEmail(
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String?,
        @Query("payload") payload: String?,
        @Body body: ResendAccountConfirmationEmailBody
    ): Call<Void>


    // Emails
    //
    // OAuth 2.0

    @POST("/oauth2/user/resend_confirmation_link")
    fun oauthResendAccountConfirmationEmail(
        @Query("client_id") clientId: Int,
        @Query("redirect_uri") redirectUri: String?,
        @Query("state") state: String,
        @Body body: ResendAccountConfirmationEmailBody
    ): Call<Void>

    //----------     Password     ----------

    // Password
    //
    // Resetting

    //ToDo:: Confirm password reset

    @POST("/api/password/reset/request")
    fun resetPassword(
        @Query("projectId") projectId: String,
        @Query("login_url") loginUrl: String?,
        @Body resetPasswordBody: ResetPasswordBody
    ): Call<Void>

    //----------     Linking Accounts     ----------


    // Linking Accounts
    //
    // Linking


    @POST("api/users/account/code")
    fun createCodeForLinkingAccounts(
        @Header("authorization") authHeader: String
    ): Call<CreateCodeForLinkingAccountResponse>

    //ToDo:: Link accounts by code

    //----------     Attributes     ----------


    // Attributes
    //
    // Client

    @POST("api/attributes/users/me/get")
    fun getUsersAttributesFromClient(
        @Header("authorization") authHeader: String,
        @Body getUsersAttributesFromClientRequest: GetUsersAttributesFromClientRequest
    ): Call<List<UserAttribute>>

    @POST("api/attributes/users/me/get_read_only")
    fun getUsersReadOnlyAttributesFromClient(
        @Header("authorization") authHeader: String,
        @Body getUsersAttributesFromClientRequest: GetUsersAttributesFromClientRequest
    ): Call<List<UserAttribute>>

    @POST("api/attributes/users/me/update")
    fun updateUsersAttributesFromClient(
        @Header("authorization") authHeader: String,
        @Body updateUsersAttributesFromClientRequest: UpdateUsersAttributesFromClientRequest
    ): Call<Void>

    //----------     User Account     ----------


    // User Account
    //
    // Devices


    @GET("/api/users/me/devices")
    fun getUsersDevices(
        @Header("authorization") authHeader: String
    ): Call<List<UsersDevicesResponse>>

    @POST("/api/users/me/devices/{device_type}")
    fun linkDeviceToAccount(
        @Header("authorization") authHeader: String,
        @Path("device_type") deviceType: String,
        @Body linkDeviceToAccountBody: AuthViaDeviceIdBody
    ): Call<Void>

    @DELETE("/api/users/me/devices/{id}")
    fun unlinkDeviceFromAccount(
        @Header("authorization") authHeader: String,
        @Path("id") id: Int
    ): Call<Void>

    // User Account
    //
    // MFA

    //User Account
    //
    // User Profile


    @POST("api/users/age/check")
    fun checkUserAge(@Body body: CheckUserAgeBody): Call<CheckUserAgeResponse>

    @GET("api/users/me")
    fun getCurrentUserDetails(@Header("authorization") authHeader: String): Call<UserDetailsResponse>

    @PATCH("api/users/me")
    fun updateCurrentUserDetails(
        @Header("authorization") authHeader: String,
        @Body updateUserDetailsBody: UpdateUserDetailsBody
    ): Call<Void>

    @GET("api/users/me/email")
    fun getCurrentUserEmail(@Header("authorization") authHeader: String): Call<EmailResponse>

    @POST("/api/users/me/link_email_password")
    fun linkEmailPassword(
        @Header("authorization") authHeader: String,
        @Query("login_url") loginUrl: String,
        @Body linkEmailPasswordBody: LinkEmailPasswordBody
    ): Call<LinkEmailPasswordResponse>

    @GET("api/users/me/phone")
    fun getUserPhone(@Header("authorization") authHeader: String): Call<PhoneResponse>

    @POST("api/users/me/phone")
    fun updateUserPhone(
        @Header("authorization") authHeader: String,
        @Body updateUserPhoneBody: UpdateUserPhoneBody
    ): Call<Void>

    @DELETE("api/users/me/phone/{phoneNumber}")
    fun deleteUserPhone(
        @Header("authorization") authHeader: String,
        @Path("phoneNumber") phoneNumber: String
    ): Call<Void>

    @DELETE("api/users/me/picture")
    fun deleteUserPicture(@Header("authorization") authHeader: String): Call<Void>

    @Multipart
    @POST("api/users/me/picture")
    fun uploadUserPicture(
        @Header("authorization") authHeader: String,
        @Part picture: MultipartBody.Part
    ): Call<PictureResponse>

     // User Account
     //
     // User Friends

    @GET("api/users/me/relationships")
    fun getUserFriends(
        @Header("authorization") authHeader: String,
        @Query("after") after: String?,
        @Query("limit") limit: Int,
        @Query("type") userFriendsRequestType: String,
        @Query("sort_by") userFriendsRequestSortBy: String,
        @Query("sort_order") userFriendsRequestSortOrder: String
    ): Call<UserFriendsResponse>

    @POST("api/users/me/relationships")
    fun updateFriends(
        @Header("authorization") authHeader: String,
        @Body updateUserFriendsRequest: UpdateUserFriendsRequest
    ): Call<Void>

    @GET("/api/users/me/social_friends")
    fun getSocialFriends(
        @Header("authorization") authHeader: String,
        @Query("platform") platform: String?,
        @Query("offset") offset: Int,
        @Query("limit") limit: Int,
        @Query("with_xl_uid") fromGameOnly: Boolean
    ): Call<SocialFriendsResponse>

    @POST("api/users/me/social_friends/update")
    fun updateSocialFriends(
        @Header("authorization") authHeader: String,
        @Query("platform") platform: String?
    ): Call<Void>

    @GET("api/users/search/by_nickname")
    fun searchUsersByNickname(
        @Header("authorization") authHeader: String,
        @Query("nickname") nickname: String,
        @Query("offset") offset: Int,
        @Query("limit") limit: Int
    ): Call<SearchUsersByNicknameResponse>

    @GET("api/users/{userId}/public")
    fun getUserPublicInfo(
        @Header("authorization") authHeader: String,
        @Path("userId") userId: String
    ): Call<UserPublicInfoResponse>

     // User Account
     //
     // Social Networks

    @GET("/users/me/login_urls")
    fun getLinksForSocialAuth(
        @Header("authorization") authHeader: String,
        @Query("locale") locale: String
    ): Call<LinksForSocialAuthResponse>

    @GET("api/users/me/social_providers")
    fun getLinkedSocialNetworks(
        @Header("authorization") authHeader: String
    ): Call<List<LinkedSocialNetworkResponse>>

    @GET("/api/users/me/social_providers/{providerName}/login_url")
    fun getUrlToLinkSocialNetworkToAccount(
        @Header("authorization") authHeader: String,
        @Path("providerName") providerName: String,
        @Query("login_url") urlRedirect: String
    ): Call<String>

    @DELETE("/api/users/me/social_providers/{providerName}")
    fun unlinkSocialNetwork(
        @Header("authorization") authHeader: String,
        @Path("providerName") providerName: String
    ): Call<Void>

    @FormUrlEncoded
    @POST("/api/oauth2/token")
    fun oauthRefreshToken(
        @Field("refresh_token") refreshToken: String,
        @Field("grant_type") grantType: String,
        @Field("client_id") clientId: Int,
        @Field("redirect_uri") redirectUri: String
    ): Call<OauthAuthResponse>

    @POST("/api/oauth2/social/mobile/{providerName}/login_with_code")
    fun oauthGetCodeBySocialCode(
        @Path("providerName") providerName: String,
        @Query("client_id") clientId: Int,
        @Query("state") state: String,
        @Query("redirect_uri") redirectUri: String,
        @Query("response_type") responseType: String,
        @Query("scope") scope: String,
        @Body authUserSocialWithCodeBody: AuthUserSocialWithCodeBody
    ): Call<OauthGetCodeResponse>
}